src/ScheduleFunctions.cpp

```diff
@@ -763,225 +738,32 @@ class IsRealizedInStmt : public IRVisitor {
-            if (!function_is_already_realized_in_stmt(func, body) &&
-                (function_is_used_in_stmt(func, body) || is_output)) {
-                body = build_pipeline(body);
-            }
-            found_compute_level = true;
-        }
-
-        if (store_level.match(for_loop->name)) {
-            debug(3) << "Found store level\n";
-            internal_assert(found_compute_level)
-                << "The compute loop level was not found within the store loop level!\n";
-
-            if (!function_is_already_realized_in_stmt(func, body) &&
-                (function_is_used_in_stmt(func, body) || is_output)) {
-                body = build_realize(body);
-            }
-
-            found_store_level = true;
-        }
-
-        // Reinstate the let statements
-        for (size_t i = lets.size(); i > 0; i--) {
-            body = LetStmt::make(lets[i - 1].first, lets[i - 1].second, body);
-        }
-
-        // Reinstate the placeholder prefetches
-        for (size_t i = placeholder_prefetches.size(); i > 0; i--) {
-            body = Prefetch::make(std::get<0>(placeholder_prefetches[i - 1]),
-                                  std::get<1>(placeholder_prefetches[i - 1]),
-                                  Region(),
-                                  std::get<2>(placeholder_prefetches[i - 1]),
-                                  const_true(),
-                                  body);
-        }
-
-        if (body.same_as(for_loop->body)) {
-            return for_loop;
-        } else {
-            return For::make(for_loop->name,
-                             for_loop->min,
-                             for_loop->extent,
-                             for_loop->for_type,
-                             for_loop->device_api,
-                             body);
-        }
-    }
-
-    // If we're an inline update or extern, we may need to inject a realization here
-    Stmt visit(const Provide *op) override {
-        if (op->name != func.name() &&
-            !func.is_pure() &&
-            func.schedule().compute_level().is_inlined() &&
-            function_is_used_in_stmt(func, op)) {
-
-            // Prefix all calls to func in op
-            Stmt stmt = build_realize(build_pipeline(op));
-            found_store_level = found_compute_level = true;
-            return stmt;
-        } else {
-            return op;
-        }
-    }
-};
-
-std::ostream& operator<<(std::ostream& out, const std::vector<Function>& v) {
-    out << "{ ";
-    for (size_t i = 0; i < v.size(); ++i) {
-        out << v[i].name();
-        if (i != v.size() - 1) {
-            out << ", ";
-        }
-    }
-    out << " }";
-    return out;
-}
-
 class InjectStmt : public IRMutator2 {
 public:
     Stmt injected_stmt;
     bool found_level;
     LoopLevel level;
 
-    InjectStmt(const Stmt &s, const LoopLevel &level)
-        : injected_stmt(s), found_level(false), level(level) {}
+    InjectStmt(Stmt s, LoopLevel level)
```

Comment:
A meta-comment here: while appropriate use of `std::move` is always a good thing, it's not clear to me that converting from passing by const-ref, to passing-by-value-with-move, is a net win. Is there any evidence that supports either as a preferred call style?

Reply:
Yes, because there's a copy made here anyway. Changing the signature to pass-by-value, then move, shows the caller that there is a copy made and that ownership is not shared.

Reply:
Does this class even need to be making a copy though? It may be better to change the members to be const refs.

Reply:
That is also a good solution. I kept the copy because the existing code copied and I didn't want to change the behavior.

Reply:
Switched to const& fields.